database/merkledb/branch_node.go

```diff
@@ -0,0 +1,384 @@
+package merkledb
+
+import (
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/database"
+)
+
+// BranchNode represents a Node with an array of Nodes and a SharedAddress
+// SharedAddress is the shared prefix of all keys under this node
+// Nodes are Addresses of other Nodes, which their addresses are suffixed by SharedAddress
+//
+// LASTPOSition is always a LeafNode and it represents the case where the LeafNode has the same Address
+// as the BranchNode SharedAddress
+//
+// Hashes are hashes of child Nodes
+// [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, a, b, c, d, f, LASTPOS ]
+//     [ x ] -> LeafNode
+//     [ not found ] -> EmptyNode
+//     [ LASTPOS ] -> LeafNode
+//
+type BranchNode struct {
+	Nodes              [UnitSize][]byte `serialize:"true"`
+	SharedAddress      Key              `serialize:"true"`
+	StoredHash         []byte           `serialize:"true"`
+	Refs               int32            `serialize:"true"`
+	previousStoredHash []byte
+	parent             Node
+	persistence        Persistence
+}
+
+// NewBranchNode returns a new BranchNode
+func NewBranchNode(sharedAddress Key, parent Node, persistence Persistence) Node {
+	return &BranchNode{
+		SharedAddress: sharedAddress,
+		Nodes:         [UnitSize][]byte{},
+		parent:        parent,
+		persistence:   persistence,
+	}
+}
+
+// GetChild returns a child node
+// returns Node - one of it's children
+// returns EmptyNode - no node in this position
+//
+func (b *BranchNode) GetChild(key Key) (Node, error) {
+	if !key.ContainsPrefix(b.SharedAddress) {
+		return NewEmptyNode(b, key), nil
+	}
+
+	// if the node CAN exist in this prefix but doesn't return an EmptyNode
+	// Node Key: AABBCC
+	// SharedAddress: AABBC
+	// FirstNonPrefix:  AABBCC - AABBC = C
+	nodeHash := b.Nodes[FirstNonPrefix(b.SharedAddress, key)]
+	if len(nodeHash) == 0 {
+		return NewEmptyNode(b, key), nil
+	}
+
+	node, err := b.persistence.GetNodeByHash(nodeHash)
+	if err != nil {
+		return nil, err
+	}
+	node.SetParent(b)
+
+	return node, nil
+}
+
+// GetNextNode returns the next node in increasing key order
+// returns Node - one of it's children
+func (b *BranchNode) GetNextNode(prefix Key, start Key, key Key) (Node, error) {
```

Comment:
Can you add a comment saying what `prefix` and `start` are?

Reply:
I think we can rewrite this to just use the greatest value between prefix, start, and key 🤔 

For example `Give me all the nodes that have the prefix with ABB, starting at ABB123, where the key is Greater than ABC`

This can be re-writen as `Give me all the nodes that have a key greater than ABC`

Is there a scenario I'm not seeing 🤔